### PR TITLE
sdk/trace: use slices.Grow() to avoid excessive runtime.growslice()

### PR DESCRIPTION
**Pull request by lizthegrey**
Profiling of our otel load generator shows excessive time in growslice. This indicates that we need to expand cap(s.attributes) all at once, rather than one element at a time in the loop.

image

```
$ ~/hny/go/bin/benchstat /tmp/old /tmp/new
goos: linux
goarch: amd64
pkg: go.opentelemetry.io/otel/sdk/trace
cpu: AMD Ryzen 5 5600X 6-Core Processor             
                                              │  /tmp/old   │              /tmp/new               │
                                              │   sec/op    │   sec/op     vs base                │
RecordingSpanSetAttributes/WithLimit/false-12   3.850µ ± 0%   2.721µ ± 2%  -29.32% (p=0.000 n=20)
RecordingSpanSetAttributes/WithLimit/true-12    8.040µ ± 1%   7.362µ ± 0%   -8.44% (p=0.000 n=20)
geomean                                         5.563µ        4.476µ       -19.55%

                                              │   /tmp/old    │               /tmp/new               │
                                              │     B/op      │     B/op      vs base                │
RecordingSpanSetAttributes/WithLimit/false-12   16.406Ki ± 0%   6.844Ki ± 0%  -58.29% (p=0.000 n=20)
RecordingSpanSetAttributes/WithLimit/true-12     15.81Ki ± 0%   11.00Ki ± 0%  -30.43% (p=0.000 n=20)
geomean                                          16.11Ki        8.677Ki       -46.13%

                                              │  /tmp/old   │              /tmp/new              │
                                              │  allocs/op  │ allocs/op   vs base                │
RecordingSpanSetAttributes/WithLimit/false-12   13.000 ± 0%   6.000 ± 0%  -53.85% (p=0.000 n=20)
RecordingSpanSetAttributes/WithLimit/true-12     18.00 ± 0%   12.00 ± 0%  -33.33% (p=0.000 n=20)
geomean                                          15.30        8.485       -44.53%
```

### CODE REVIEW

File: sdk/trace/span.go [hunk 1]
```diff
@@ -242,6 +244,7 @@ func (s *recordingSpan) SetAttributes(attributes ...attribute.KeyValue) {
 
 	// Otherwise, add without deduplication. When attributes are read they
 	// will be deduplicated, optimizing the operation.
+	s.attributes = slices.Grow(s.attributes, len(attributes))
```
Comment (pellared) [inline sdk/trace/span.go:247]: How about inlining the `slices.Grow` implementation instead of referencing `golang.org/x/exp`?

**Review by pellared — COMMENTED** [review]
Do we have some benchmarks that would show the reduced number of allocations? Can you add  a [benchstat](https://pkg.go.dev/golang.org/x/perf/cmd/benchstat) result in the description?

File: sdk/trace/span.go (hunk 1 above)
Comment (lizthegrey) [reply sdk/trace/span.go:247]: Certainly happy to inline it if that's easier.

Comment (pellared) [reply sdk/trace/span.go:247]: Right now it is used only in `sdk/trace/span.go` right? I think we can add a comment and issue that we should use https://pkg.go.dev/slices#Grow instead once we stop supporting Go 1.20.

Comment (lizthegrey) [reply sdk/trace/span.go]: Correct, after Go1.20 is no longer supported we can use slices#Grow instead. I've inlined as requested.

**Comment by lizthegrey** [comment]
Interesting, adding the bench shows no change in allocs, which is very suspicious. I looked at my bench and it looks like I am calling SetAttributes() one at a time, which isn't otel's fault, it's my fault, I'll fix my loadgen binary.

**Comment by lizthegrey** [comment]
> Interesting, adding the bench shows no change in allocs, which is very suspicious. I looked at my bench and it looks like I am calling SetAttributes() one at a time, which isn't otel's fault, it's my fault, I'll fix my loadgen binary.

That's because I introduced a bug while inlining. It's fixed now. Description updated.

File: sdk/trace/span_test.go [hunk 2]
```diff
@@ -243,3 +244,20 @@ func TestTruncateAttr(t *testing.T) {
...
+func BenchmarkRecordingSpanSetAttributes(b *testing.B) {
+	tp := NewTracerProvider(WithSampler(AlwaysSample()))
+	tracer := tp.Tracer("tracer")
+	var attrs []attribute.KeyValue
+	for i := 0; i < 100; i++ {
+		attr := attribute.String(fmt.Sprintf("hello.attrib%d", i), fmt.Sprintf("goodbye.attrib%d", i))
+		attrs = append(attrs, attr)
+	}
+
+	ctx := context.Background()
```
Comment (dmathieu) [inline sdk/trace/span_test.go:257]: How about adding a `b.ResetTimer()` just before the loop?

```suggestion
	ctx := context.Background()
	b.ResetTimer()
```

**Review by dmathieu — COMMENTED** [review]
Out of curiosity, do you have access to profiling data after the change?

File: sdk/trace/span_test.go [hunk 3]
```diff
@@ -243,3 +244,20 @@ func TestTruncateAttr(t *testing.T) {
 		})
 	}
 }
+
+func BenchmarkRecordingSpanSetAttributes(b *testing.B) {
```
Comment (pellared) [inline sdk/trace/span_test.go:248]: Can you add a call to https://pkg.go.dev/testing#B.ReportAllocs?

File: sdk/trace/span.go [hunk 4]
```diff
@@ -208,6 +208,15 @@ func (s *recordingSpan) SetStatus(code codes.Code, description string) {
 	s.status = status
 }
 
+// ensureAttributesCapacity inlines functionality from golang.org/x/exp/slices.Grow
```
Comment (pellared) [inline sdk/trace/span.go:211]:
```suggestion
// ensureAttributesCapacity inlines functionality from slices.Grow
```

File: sdk/trace/span.go [hunk 5]
```diff
@@ -208,6 +208,15 @@ func (s *recordingSpan) SetStatus(code codes.Code, description string) {
...
+// so that we can avoid needing to import golang.org/x/exp for go1.20.
+// Once support for go1.20 is dropped, we can use slices.Grow in 1.21+ instead.
```
Comment (pellared) [inline sdk/trace/span.go:213]:
```suggestion
// Once support for go1.20 is dropped, we can use slices.Grow in 1.21+ instead.
// Tracking issue: https://github.com/open-telemetry/opentelemetry-go/issues/4819.
```

File: sdk/trace/span.go [hunk 6]
```diff
@@ -277,6 +287,9 @@ func (s *recordingSpan) addOverCapAttrs(limit int, attrs []attribute.KeyValue) {
 
 	// Now that s.attributes is deduplicated, adding unique attributes up to
 	// the capacity of s will not over allocate s.attributes.
+	if limit-len(s.attributes) > 0 {
```
Comment (pellared) [inline sdk/trace/span.go:290]: Should it not be the following?

```suggestion
	if len(s.attributes) > limit {
```

Do we have any benchmark that shows this improvement? How about setting a limit in `BenchmarkRecordingSpanSetAttributes`?

Comment (lizthegrey) [reply sdk/trace/span.go:290]: Oh yes, nice catch on the sense change. I will add the benchmark.

File: sdk/trace/span_test.go (hunk 3 above)
Comment (lizthegrey) [reply sdk/trace/span_test.go:248]: Will do.

**Review by lizthegrey — COMMENTED** [review]
Thanks for the feedback, I'll get that sorted out.

And yes, I have profiles before/after.

**Comment by lizthegrey** [comment]
Profiles after: it's much better, and now the problem has shifted to dedupeAttributes which is a separate map allocation problem that we can't pre-allocate our way out of because Go doesn't support map capacities yet. https://github.com/golang/go/issues/56182

Screenshot 2024-01-10 15 48 06
Screenshot 2024-01-10 15 47 53

**Comment by lizthegrey** [comment]
Benchmark updated and bug in the pre-sizing fixed!!

**Review by pellared — APPROVED** [review]

**Review by dmathieu — APPROVED** [review]

**Comment by lizthegrey** [comment]
@MadVikingGod any chance of squeezing this into v1.22? It's ready to go otherwise (2 approvals)

**Review by MrAlias — APPROVED** [review]

**Pull request merged**